### ISSUE TITLE
🚀 Feature: Permission caching

### ISSUE THREAD

**Issue by drodil** [backstage/backstage#23085]
### 🔖 Feature description

Currently, permissions are checked multiple times for the same things in a short period of time. This can get very slow with lots of entities and thus proposal is to use `CacheManager` to cache permission responses for a short time. 

### 🎤 Context

Especially search and catalog can get very slow due to permission checking with lots of data.

### ✌️ Possible Implementation

- Allow defining TTL for caching permission responses in the `permissions` configuration
- Use `CacheManager` to save and fetch permission responses per user/token

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by freben** [comment]
Do you have a particular use case in mind where this cache would be particularly effective? Or put differently, do you have a reproduction scenario to look at, where the same permission for the same resource is checked multiple times?

**Comment by drodil** [comment]
Not necessarily checked multiple times but many times in a short period. For example, when you refine the search term, the same checks are done repeatedly for probably the same resources.

**Comment by freben** [comment]
Thanks, that makes sense and helps to have as an example use case when someone tries to implement it.

**Comment by vinzscam** [comment]
Catalog permissions are translated and embedded into the same db query used for querying the entities, therefore no extra request is done. I don't think caching would help with anything here. Have you hit any specific issue with catalog?

**Comment by drodil** [comment]
The only problem is that it's slow when there are _a lot_ of entities (10000+). This shows in both search and in catalog so I would think it's because of the permissions. Searching for stuff that has no permissions attached is quite fast and the results are returned in less than a second (around 600ms) even when the index is at least 50 times bigger than with catalog entities. When searching for catalog items, it can take up to 5 seconds to get the first page of results. 

Which reminds me about PR #22594 ... Is there something that could be done for the indexes? I think all `search` table queries use both `key` and `value` and as they are currently in different indexes which might cause some delay? Not sure though, just a hunch. Should run some real benchmarking for this to find out the bottlenecks.

**Comment by drodil** [comment]
> * for search, entities are authorized result-by-result (for technical reasons related to the underlining search engines) but there is a mechanism to put a "cap" in case a request takes more than a fixed max time

Ain't this exactly the case where caching would help?

**Comment by vinzscam** [comment]
I'm curious to know if you see any difference in performance by using the changes of #23088.
I'm asking because the only benefit I see is when plugins invokes `permission.authorize()` which in case of conditional response is going to save some time. But, both catalog and search call only `permission.authorizeConditional()`, which returns a straight AUTHORIZE/DENY or a CONDITIONAL (which is relatively simple to compute). So what we are actually caching isn't something computational expensive.

**Comment by drodil** [comment]
Oh, so conditional permissions are not handled by the permission backend? If so, then yes, caching in the router won't help a lot. I will try adding a `search` table index for `key+value` in our dev environment and see if it has any benefits to the performance.

**Comment by drodil** [comment]
So I made the key value index (concurrently for postgres) and it seems it replaces the separate indexes for key and value completely. Those are not used anymore at all after creating the new index:

```
 relid  | indexrelid | schemaname | relname |        indexrelname         |  idx_scan   | idx_tup_read | idx_tup_fetch
---------+------------+------------+---------+-----------------------------+-------------+--------------+---------------
 7526820 |   61963899 | public     | search  | search_entity_key_value_idx |       23478 |     47796994 |        489751
 7526820 |    7526832 | public     | search  | search_key_idx              |    96702171 | 178602269787 |       5564382
 7526820 |    7526833 | public     | search  | search_value_idx            |   517453715 |  23912811589 |     120210880
 7526820 |    7526831 | public     | search  | search_entity_id_idx        | 15775484311 | 457025156509 |  444485542639
```

After some time:

```
 relid  | indexrelid | schemaname | relname |        indexrelname         |  idx_scan   | idx_tup_read | idx_tup_fetch
---------+------------+------------+---------+-----------------------------+-------------+--------------+---------------
 7526820 |   61963899 | public     | search  | search_entity_key_value_idx |       36365 |     48176712 |        823942
 7526820 |    7526832 | public     | search  | search_key_idx              |    96702171 | 178602269787 |       5564382
 7526820 |    7526833 | public     | search  | search_value_idx            |   517453715 |  23912811589 |     120210880
 7526820 |    7526831 | public     | search  | search_entity_id_idx        | 15775718891 | 457031167027 |  444490776391
```

Additionally, the catalog feels much, much snappier. Filtering 1000+ entities in the catalog page dropped from ~2 seconds to under 500ms. 

I think I will apply this same index for the rest of our environments and report more specific findings here.

**Comment by drodil** [comment]
So applied the index to another instance. Here are the results:

# Catalog search key+value index

Query plan query used:

```sql
SELECT * FROM final_entities LEFT JOIN search ON final_entities.entity_id = search.entity_id WHERE search.key = 'kind' and search.value = 'component' AND final_entities.final_entity IS NOT NULL;
```

## Before indexing

Before adding a new index with both key and value, the results are as follows.

### Query plan

```
                                                 QUERY PLAN
------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=660.48..2073.82 rows=126 width=1306)
   ->  Bitmap Heap Scan on search  (cost=660.06..1139.63 rows=126 width=118)
         Recheck Cond: (((value)::text = 'component'::text) AND ((key)::text = 'kind'::text))
         ->  BitmapAnd  (cost=660.06..660.06 rows=126 width=0)
               ->  Bitmap Index Scan on search_value_idx  (cost=0.00..249.42 rows=8115 width=0)
                     Index Cond: ((value)::text = 'component'::text)
               ->  Bitmap Index Scan on search_key_idx  (cost=0.00..410.33 rows=18921 width=0)
                     Index Cond: ((key)::text = 'kind'::text)
   ->  Index Scan using final_entities_entity_id_idx on final_entities  (cost=0.41..7.41 rows=1 width=1188)
         Index Cond: (entity_id = search.entity_id)
         Filter: (final_entity IS NOT NULL)
(11 rows)
```

### UI Testing

| Request                                            | Total time | Entities loaded |
|----------------------------------------------------|------------|-----------------|
| Initial request with kind filter (component)       | 14.08s     | 7951            |
| Request with single type + kind filter             | 6.38s      | 6448            |
| Request with single other type + kind filter       | 1.07s      | 449             |
| Request with single annotation value + kind filter | 835ms      | 11              | 
| Change kind filter to another kind (system)        | 1.89s      | 1156            |

(Repository, Pala business, ASUPA, )

## After indexing

Creating index with:

```sql
CREATE INDEX CONCURRENTLY search_key_value_idx ON search (key, value);
```

### Query plan

```
                                                QUERY PLAN
------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=6.50..1666.66 rows=150 width=1306)
   ->  Bitmap Heap Scan on search  (cost=6.09..574.15 rows=150 width=118)
         Recheck Cond: (((key)::text = 'kind'::text) AND ((value)::text = 'component'::text))
         ->  Bitmap Index Scan on search_key_value_idx  (cost=0.00..6.05 rows=150 width=0)
               Index Cond: (((key)::text = 'kind'::text) AND ((value)::text = 'component'::text))
   ->  Index Scan using final_entities_entity_id_idx on final_entities  (cost=0.41..7.28 rows=1 width=1188)
         Index Cond: (entity_id = search.entity_id)
         Filter: (final_entity IS NOT NULL)
(8 rows)
```

### UI Testing


| Request                                            | Total time | Entities loaded |
|----------------------------------------------------|------------|-----------------|
| Initial request with kind filter (component)       | 5.39s      | 7951            |
| Request with single type + kind filter             | 2.33s      | 6448            |
| Request with single other type + kind filter       | 283ms      | 449             |
| Request with single annotation value + kind filter | 181ms      | 11              | 
| Change kind filter to another kind (system)        | 1.33s      | 1156            |

**Comment by freben** [comment]
@drodil Were you meaning to put those comments on #22594 or somewhere like that? Maybe I'm misunderstanding

**Comment by drodil** [comment]
@freben yeah those could be added there, a bit wrong issue maybe 😆

**Comment by vinzscam** [comment]
Closed by #22594
